aws-api/src/main/java/com/amplifyframework/api/aws/auth/ApiRequestSigner.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.api.aws.auth;
+
+import androidx.annotation.NonNull;
+
+import com.amplifyframework.util.Empty;
+
+import com.amazonaws.DefaultRequest;
+import com.amazonaws.http.HttpMethodName;
+import com.amazonaws.util.IOUtils;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLDecoder;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import okhttp3.MediaType;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okio.Buffer;
+
+/**
+ * Abstract class that defines some basic functionality with regards to signing AppSync requests.
+ * Implementations of this class should implement the {@link ApiRequestSigner#addAuthHeader(com.amazonaws.Request)}
+ * method.
+ */
+public abstract class ApiRequestSigner {
```

Comment:
It seems like the `JWTTokenApiRequestSigner` instance of this type isn't actually a signer.

Maybe the base type should just be:

```java
public interface RequestDecorator {
    public <T> Request<T> decorate(Request<T> request);
}
```

Then, the signing stuff can be a final (closed) utility class. I don't think you need inheritance.

Each individual implementation of the request decorator could get the signing values from the signer, and then populate them into the header using their specific header keys.

Reply:
I like that. As you mentioned, there's a bit of an identity crisis with the sigv4-related code. It probably makes sense to untangle some that.